### PR TITLE
Cache and restore solutions on XFEM-affected nodes/elems

### PR DESCRIPTION
**Pull request by bwspenc**
ref #8763

### CODE REVIEW

File: framework/include/base/XFEMInterface.h
Comment (permcody) [inline framework/include/base/XFEMInterface.h]: `_moose_mesh2`? This isn't a very descriptive variable name. Why are there two? Can you name it something better?

Unrelated, we prefer `nullptr` over `NULL` in C++11 because of places where we infer type. NULL is actually an `int`, feel free to replace `NULL` when you see it, and avoid adding new ones.

File: modules/xfem/src/base/XFEM.C
Comment (permcody) [inline modules/xfem/src/base/XFEM.C]: Ugh, this probably should be moved into libMesh at some point.

File: modules/xfem/src/base/XFEM.C
Comment (permcody) [inline modules/xfem/src/base/XFEM.C]: What's with the variable names here? They don't follow code standards.

File: modules/xfem/src/base/XFEM.C
Comment (permcody) [inline modules/xfem/src/base/XFEM.C]: Wait, are you sure about this? Your processor may not own some of the nodes of a semilocal element. Sure the solution will be correct but do you attempt to write these values later?

File: modules/xfem/src/base/XFEM.C
Comment (permcody) [inline modules/xfem/src/base/XFEM.C]: Just so it's on your radar. This line of code will fail badly if we attempt to run with DistributedMesh. I know that's not your goal here but at least get in the habit of using mooseAssert so that when we try we don't have to dig in with the debugger to find all these problems.

File: modules/xfem/src/base/XFEM.C
Comment (permcody) [inline modules/xfem/src/base/XFEM.C]: Hmm... This may not be correct. isSemiLocal just means that the element is attached to a local element. libMesh has better methods for testing for ownership, ghosting, evaluable, these days.

File: modules/xfem/src/base/XFEM.C
Comment (permcody) [inline modules/xfem/src/base/XFEM.C]: There is a lot of code duplication in these next four methods. Have you considered breaking this into a few more methods and dropping all of the duplicate code?

File: modules/xfem/src/base/XFEM.C
Comment (permcody) [inline modules/xfem/src/base/XFEM.C]: This is pretty special logic here and isn't documented. It looks like you are doubling storage so you can store current and old values. However you don't say the order in which you store this information. Also, why not store older values? Will that ever be necessary?

File: modules/xfem/src/base/XFEM.C
Comment (permcody) [inline modules/xfem/src/base/XFEM.C]: Alternative self documenting data structure below. I'm fine with either but you should at least mention what you are doing.

```
struct DofValues
{
  Real current;
  Real old;
  Real older;
}

std::map<unique_id_type, DofValues> data;
```

File: modules/xfem/src/base/XFEM.C
Comment (permcody) [inline modules/xfem/src/base/XFEM.C]: So why is the scratch vector a class member instead of a local member? You are reallocating and copying it around anyway. In fact, you would be better off with using `::assign` here to avoid reallocation here if that's the optimization you are after.

File: modules/xfem/src/base/XFEM.C
Comment (permcody) [inline modules/xfem/src/base/XFEM.C]: Yeah, I see it now. There's a little voodoo on this line with your storage indexing. It probably should be documented.

**Review by permcody — COMMENTED** [review]
I had several comments. I know this is a bug fix but it seems like there's a lot of quickly developed code here which could be cleaned up and made more general, faster, etc. I'll let you look it over. I don't see anything that prevents merging but there are some maintenance concerns with this new code.

**Comment by bwspenc** [comment]
@permcody All good points that I'll address.  This does need some more cleanup -- I wanted some initial feedback on the concept here (and it looks like the tests didn't even pass in my first attempt -- oops!)

I'll summarize the problem: We only add new nodes with XFEM, and they always copy their state from an existing nodes. I previously stored the old node to copy the solution from, but I found that in some cases, the old nodes also ended up with new DOF indices, so their solution values were lost.  It seems like the most reliable thing to do is just cache the solution and then restore it, which ended up being a lot more work than I initially anticipated.  What do you think -- Is that a crazy thing to do? It works reliably now, even on some parallel cases where the crack is running along right next to a processor boundary where it failed before.

Another couple of questions: It looks like the variables can also have components.  Do we ever use more than 1 component in MOOSE?  I'm only caching/recovering the 1st component.

Also, how about the older solutions.  I'm caching/recovering the current and old solution for nl and aux.  Do both of these actually have an older solution?  It looks like aux does, but I'm not sure about nl.  We never use them, so I don't really care about them, but I should probably copy those as well if they're there just so this doesn't unexpectedly break something.

**Comment by permcody** [comment]
We do use more than one component in MOOSE for other shape function families: Higher order monomials, hierarchical?, maybe others?

Older solutions do exist for both aux and nl. They are populated and used anytime somebody requests access to those older solutions on either system respectively. 

The reason I picked on you is that I've already been down this cache and restore path for my work. Made all the normal mistakes, fixed many, but there are still a few assumptions, etc.:
See my version here:
https://github.com/idaholab/moose/blob/devel/modules/phase_field/src/postprocessors/GrainTracker.C#L1358-L1436

I also only grab the first component in my implementation as you can see. Mine isn't even as general as yours in that I didn't bother to implement an elemental version. Also, looking at my code, I recall that setting an old or older solution on a steady state solve is bad (like seg fault bad). 

All of this caching code really should be put in the framework eventually so we don't end up writing it yet again for some other task.

**Comment by bwspenc** [comment]
@permcody Thanks for the info. If you've done this another place, I guess I'm not crazy.  It's going to be next week before I can get back to this, but I'm happy to clean this up then before you merge this.  @jiangwen84 you can just grab this commit to use for the milestone work in the mean-time.

File: modules/xfem/src/base/XFEM.C
Comment (bwspenc) [inline modules/xfem/src/base/XFEM.C]: The semilocal element might be attached to local nodes.  We err on the side of storing more data than we actually need here, and then only do the setting of the solution on entities that are locally owned.  This is necessary for this to work in cases where the newly-created nodes end up on different processors.

**Comment by bwspenc** [comment]
@permcody I've significantly re-worked this and I believe that I've addressed your comments.  I added a test that confirms that this works for both nodal and element variables, including ones with multiple components.  Without this fix, this test fails in parallel in multiple ways.

**Review by permcody — APPROVED** [review]
This looks better. At some point in the future we'll still have to factor out all this solution update logic though. Not today...

**Pull request merged**
